### PR TITLE
Replace current regex engine with PCRE2

### LINKED ISSUE #3730: Regex is slow
### Work environment

| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | Debian x86_64
| File format of the file you reverse (mandatory)      | -
| Architecture/bits of the file (mandatory)            | -
| `rizin -v` full output, **not truncated** (mandatory)  | rizin 0.7.0 @ linux-x86-64 commit: 7febd38e9d00a8b97b32f974d87aa7a9af37dd39

### Expected behavior

`rz_regex_exec()` should be roughly as fast as `regexec` (from `<regex.h>`).  

### Actual behavior

`rz_regex_exec` performance

perf

`regexec` performance is at least 20-40 times faster.

perf2

### Steps to reproduce the behavior

Performance is measured like this:

```
// Checkout commit which supports TriCore tokens
git checkout 7febd38e9d00a8b97b32f974d87aa7a9af37dd39
// Build rizin normally
valgrind --tool=callgrind rizin -Qc "pd 50" test/bins/tricore/Blinky_LED_1_KIT_TC367_TFT.elf
// Build rizin with https://github.com/Rot127/rizin/tree/regex-perf
valgrind --tool=callgrind rizin -Qc "pd 50" test/bins/tricore/Blinky_LED_1_KIT_TC367_TFT.elf
// Compare performance of regexec/rz_regex_exec in rz_asm_tokenize_asm_regex()
kcachegrind callgrind.out.*
```

### Additional Logs, screenshots, source code,  configuration dump, ...

cc @GustavoLCR

### PR DESCRIPTION
**Pull request by Rot127**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Replace the current regex engine with PCRE2.

**Test plan**

All green.

**Closing issues**

closes https://github.com/rizinorg/rizin/issues/3730

Partially addresses https://github.com/rizinorg/rizin/issues/4055

**Todo**

- [x] Add a `rz_vector_pop_ptr()` function which doesn't use `memcpy()` and use it in `match_all_flat()`. Or add a `rz_vector_concat()` function with the same functionality.
- [x] Terminology for groups and matches is weird.
- [x] docs
- [ ] Open bug report for tcc and pcre2 compilation and `__asm`
- [x] Test `rz_regex_get_match_name`

### CODE REVIEW

**Review by XVilka — APPROVED** [review]
LGTM if all green

**Review by wargio — APPROVED** [review]

**Review by XVilka — CHANGES_REQUESTED** [review]
Cross-build fails:
```
librz/util/meson.build:172:19: ERROR: Tried to mix libraries for machines build machine and host machine in target 'rz_util_native' This is not possible in a cross build.
```
 Just define native dependency for cross-build separately, like other libs in the same file

File: librz/cons/less.c [hunk 1]
```diff
@@ -3,9 +3,10 @@
 // SPDX-License-Identifier: LGPL-3.0-only
 
 #include <rz_cons.h>
-#include <rz_regex.h>
+#include <rz_util/rz_regex.h>
 #include <rz_util.h>
```
Comment (wargio) [inline librz/cons/less.c:7]: `#include <rz_util/rz_regex.h>` should be added in `#include <rz_util.h>`

File: librz/cons/less.c [hunk 2]
```diff
@@ -3,9 +3,10 @@
...
 #include "pager_private.h"
+#include "rz_vector.h"
```
Comment (wargio) [inline librz/cons/less.c:9]: also this is global not local.

**Comment by arrowd** [comment]
Why use static linking here? I'm getting
```
ld: error: relocation R_X86_64_32S cannot be used against symbol '_pcre2_ucd_records_8'; recompile with -fPIC
```
on FreeBSD when building against system pcre2.

**Comment by wargio** [comment]
@Rot127

**Comment by Rot127** [comment]
Thanks for the report! Will fix it in two days.

If you need a quick fix, you can add the C flag in  `subprojects/packagefiles/pcre2/meson.build`.
And run `git clean -dxff subprojects/pcre2*` afterwards.

**Pull request merged**